packages/flutter_tools/lib/src/compile.dart

```diff
@@ -1064,6 +1060,38 @@ class DefaultResidentCompiler implements ResidentCompiler {
     server.kill();
     return server.exitCode;
   }
+
+  Future<void> _writelnToServerStdin(String line, {
+    bool printTrace = false,
+  }) async {
+    await _writelnToServerStdinAll(<String>[line], printTrace: printTrace);
+  }
+
+  // TODO(andrewkolos): Concurrent calls to ProcessUtils.writelnToStdinUnsafe
+  //  against the same stdin will result in an exception. To guard against this,
+  //  we need to force calls to run serially. Ideally, this wouldn't be
+  //  necessary since we shouldn't have multiple concurrent writes to the
+  //  compiler process. However, we do (https://github.com/flutter/flutter/issues/152577).
```

Comment:
I'm okay with not shaving this yak now, because 1) event loop turns and stdin flushes are pretty fast, so the risk of it changing any existing races probably isn't high and 2) this is (potentially) blocking us getting stack traces on the top crasher of the flutter tool, which I want to resolve ASAP.

Reply:
Side note: maybe instead of `ProcessUtils.writelnToStdinUnsafe` and friends, we should create an abstraction (say, `ProcessStdinWriter`) that manages this stuff for us. While we are at it, maybe we should just replace `Process` from `dart:io` with our own wrapper entirely 🙃.